### PR TITLE
Thrift include support

### PR DESCRIPTION
**Pull request by malandrew**
This PR implements thrift include support with the ability to name the include.

Tests cover include and possible errors.

### CODE REVIEW

File: thrift.js [hunk 1]
```diff
@@ -123,17 +162,58 @@ Thrift.prototype._definitionProcessors = {
...
+        if (definitionProcessor) {
+            self[definitionProcessor](def);
         }
     }
 };
 
+Thrift.prototype.compileInclude = function compileInclude(def) {
+    var self = this;
+
+    if (def.id.indexOf('./') === 0 || def.id.indexOf('../') === 0) {
```
Comment (kriskowal) [inline thrift.js]: `id.lastIndexOf('./', 0) === 0)` for quicker fail

File: thrift.js [hunk 2]
```diff
@@ -228,13 +308,30 @@ Thrift.prototype.resolve = function resolve(def) {
...
     } else if (def.type === 'Identifier') {
-        if (!self.types[def.name]) {
-            err = new Error('cannot resolve reference to ' + def.name + ' at ' + def.line + ':' + def.column);
-            err.line = def.line;
-            err.column = def.column;
-            throw err;
+        var nameParts = def.name.split('.');
+        var types = self.types;
+
+        for (var i = 0, len = nameParts.length; i < len; i++) {
```
Comment (kriskowal) [inline thrift.js]: this is an unnecessary micro-op. we don’t do this for v8-only code.

File: thrift.js [hunk 3]
```diff
@@ -228,13 +308,30 @@ Thrift.prototype.resolve = function resolve(def) {
...
+            var id = nameParts.shift();
```
Comment (kriskowal) [inline thrift.js]: should not be necessary to mutate

File: thrift.js [hunk 4]
```diff
@@ -228,13 +308,30 @@ Thrift.prototype.resolve = function resolve(def) {
...
+                }
+                return types[id].link(self);
+            } else {
+                if (!self.modulesByName[id]) {
+                    err = new Error('cannot resolve module reference ' + id + ' for ' + def.name + ' at ' + def.line + ':' + def.column);
+                    err.line = def.line;
+                    err.column = def.column;
+                    throw err;
+                }
+                types = self.modulesByName[id].types;
```
Comment (kriskowal) [inline thrift.js]: let's break the resolve-identifier code into a small function

File: thrift.js (hunk 1 above)
Comment (malandrew) [inline thrift.js]: huh?  

```
$ node
> '../foo/bar/'.lastIndexOf('./', 0) === 0
false
> './foo/bar/'.lastIndexOf('./', 0) === 0
true
```

File: thrift.js (hunk 2 above)
Comment (malandrew) [inline thrift.js]: You're talking about the length caching right?

File: thrift.js (hunk 3 above)
Comment (malandrew) [inline thrift.js]: you mean the `Array.shift()` approach? It just seemed easier to implement and had no other uses for nameParts. I can change this to just looking at elements by index. Out of curiosity, what are the benefits of looking up by index over mutation? Performance?

File: thrift.js (hunk 4 above)
Comment (malandrew) [inline thrift.js]: Cool deal. I agree that it was a bit much for this function.

File: thrift.js [hunk 5]
```diff
@@ -123,17 +162,58 @@ Thrift.prototype._definitionProcessors = {
...
+            self[definitionProcessor](def);
         }
     }
 };
 
+Thrift.prototype.compileInclude = function compileInclude(def) {
+    var self = this;
+
+    if (def.id.indexOf('./') === 0 || def.id.indexOf('../') === 0) {
+        var thriftFile = path.resolve(self.dirname, def.id);
```
Comment (abhinav) [inline thrift.js]: What happens here when `self.dirname` is not set?

File: thrift.js [hunk 6]
```diff
@@ -123,17 +162,58 @@ Thrift.prototype._definitionProcessors = {
...
+                throw Error(
+                    'Thrift include filename is not valid thrift identifier'
+                );
+            }
+        }
+
+        self.modulesByName[ns] = self.modulesByPath[thriftFile] =
+            Thrift.loadSync({
+                thriftFile: thriftFile,
+                strict: self.strict
```
Comment (abhinav) [inline thrift.js]: `Thrift.loadSync` also links the Thrift object, right? This won't work with cyclic imports.

You probably want to do this in two steps: Compile everything first, and then link it. Compilation creates objects that can be referenced across modules, and link will actually resolve those references.

File: test/include-service.thrift [hunk 7]
```diff
@@ -0,0 +1,9 @@
+include common "./include-types.thrift"
```
Comment (abhinav) [inline test/include-service.thrift]: Oh you also implemented include-as? I would put a note somewhere in the documentation for this feature that this makes Thrift files incompatible with Apache Thrift. Also, since Python doesn't support this yet, probably mention the at-your-own-risk thing.

File: thrift.js (hunk 1 above)
Comment (kriskowal) [inline thrift.js]: both cases, but using lastIndexOf with a start index of 0 instead of indexOf is the key to avoiding a linear scan.

File: thrift.js [hunk 8]
```diff
@@ -46,6 +49,10 @@ var ThriftMap = require('./map').ThriftMap;
 var ThriftConst = require('./const').ThriftConst;
 var ThriftTypedef = require('./typedef').ThriftTypedef;
 
+var casThriftCache = {};
```
Comment (kriskowal) [inline thrift.js]: It pains me to say, this is not the right problem for a content address store. It is valid for multiple documents to have identical content, and hashing every included file is overkill. Not even necessary to canonicalize the path of the file for this case, quite sufficient to fully qualify the include path and use that as the cache key.

File: thrift.js [hunk 9]
```diff
@@ -73,6 +87,25 @@ function Thrift(options) {
...
+
+    options.source = fs.readFileSync(options.thriftFile, 'ascii');
+
+    var hash = crypto.createHash('md5')
+        .update(options.source)
+        .digest('hex');
+
+    var thrift = casThriftCache[hash] = filepathThriftCache[filepath] =
+        casThriftCache[hash] || new Thrift(options);
+
```
Comment (kriskowal) [inline thrift.js]: This is where you need to drive separate compile and link steps. Don’t rely on the Thrift constructor to collect the transitive dependencies. Drive the transitive load+compile here (since compilation is necessary to infer deps) and then compile each thrift in the resultant "cache". Also, please call it a memo, not a cache.

File: thrift.js [hunk 10]
```diff
@@ -123,17 +162,58 @@ Thrift.prototype._definitionProcessors = {
...
+        // If include isn't name, get filename sans *.thrift file extension.
+        if (!ns) {
+            var basename = path.basename(def.id);
+            ns = basename.slice(0, basename.length - 7);
+            if (!validThriftIdentifierRE.test(ns)) {
+                throw Error(
+                    'Thrift include filename is not valid thrift identifier'
+                );
+            }
+        }
```
Comment (kriskowal) [inline thrift.js]: Lets separate the import-as feature into another PR.

File: thrift.js [hunk 11]
```diff
@@ -178,6 +258,15 @@ Thrift.prototype.compileService = function compileService(def) {
...
     var service = new ThriftService({strict: self.strict});
     service.compile(def, self);
+
+    if (def.baseService) {
+        var baseService = self.resolveInclude(def, 'baseService', 'services');
+        for (var index = 0; index < baseService.functions.length; index++) {
+            var thriftFunction = baseService.functions[index];
+            service.addFunction(thriftFunction);
+        }
+    }
```
Comment (kriskowal) [inline thrift.js]: Let’s also put support for extends in a separate PR.

File: thrift.js [hunk 12]
```diff
@@ -297,4 +379,33 @@ Thrift.prototype.resolveMapConst = function resolveMapConst(def) {
     return map;
 };
 
+Thrift.prototype.resolveInclude = function resolveInclude(def, prop, defType) {
```
Comment (kriskowal) [inline thrift.js]: Let’s call this resolveIdentifier. This does not necessarily pierce an include.

File: test/include.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+});
+
+function makeThriftLoader(filename) {
+    return function thriftLoader() {
+        return Thrift.loadSync({
+            thriftFile: path.join(__dirname, filename),
+            strict: false
+        });
+    };
+}
```
Comment (kriskowal) [inline test/include.js]: Let’s not use a utility function that obscures recommended usage. We can inline this in each test without much trouble.

File: test/include.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        'throws when module name from filename is an invalid thrift identifier'
+    );
+    assert.end();
+});
+
+function makeThriftLoader(filename) {
+    return function thriftLoader() {
+        return Thrift.loadSync({
+            thriftFile: path.join(__dirname, filename),
+            strict: false
```
Comment (kriskowal) [inline test/include.js]: Also, let’s cover with strict left as true default.

File: test/include.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        /Thrift include filename is not valid thrift identifier/,
+        'throws when module name from filename is an invalid thrift identifier'
+    );
+    assert.end();
+});
+
+function makeThriftLoader(filename) {
+    return function thriftLoader() {
+        return Thrift.loadSync({
+            thriftFile: path.join(__dirname, filename),
```
Comment (kriskowal) [inline test/include.js]: Perhaps we can call this entryPoint since that will line up with the name used in the Meta::thriftIDL endpoint.

File: test/include.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    var importedThrift = makeThriftLoader('include-child.thrift')();
+
+    var typeImportedByMainThrift = mainThrift
+        .types
+        .BatchGetResponse
+        .fieldsByName
+        .items
+        .valueType
+        .rw
+        .valueType;
```
Comment (kriskowal) [inline test/include.js]: .valueType.rw.valueType should be identical to .valueType.

**Comment by kriskowal** [comment]
I see you’re using extends as a motivating case for includes. Perhaps we should extract that out in a separate PR first, or import a typedef or const for the include tests. Also the test cases should probably use the public interface. We don’t want to test for implementation details, especially ones we have to change, like the thrift model indexes.

**Comment by kriskowal** [comment]
Also, please add a test for cyclic includes to validate whether we’ve got compile and link phase separation. The simplest way to do this would be to have a file include itself.

**Comment by malandrew** [comment]
Given the comments and need to pull things out. I'm going to leave this PR as is and implement `extends` first as a PR and then rebuild this on top of that PR once it's merged in. I have two appointments this afternoon, so I'll do all this tomorrow.

File: thrift.js (hunk 10 above)
Comment (malandrew) [inline thrift.js]: Okay. I'll add this after. I wanted to add this now so i could use filename like all the others in the test folder, since the current filenaming strategy doesn't product valid thrift identifiers. That was the impetus for including this now

File: thrift.js (hunk 11 above)
Comment (malandrew) [inline thrift.js]: Done. #87

File: test/include.js (hunk 13 above)
Comment (malandrew) [inline test/include.js]: Will resolve.

File: thrift.js (hunk 11 above)
Comment (malandrew) [inline thrift.js]: Once that is landed on master, I'll rebase this PR.

File: test/include.js (hunk 16 above)
Comment (malandrew) [inline test/include.js]: Strangely, it's not. I get this error:

```
⨯ loads a thrift file with imports synchronously
  not ok 469 Type imported correctly
    ---
      operator: equal
      expected:
        { Constructor: [Function: thriftrw_Item], fieldNames: [ 'key', 'value' ], fields: [ { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} }, { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} } ], fieldsById: { 1: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} }, 2: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} } }, fieldsByName: { key: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} }, value: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: { annotations: null, baseType: 'string' }, valueType: {} } }, fullName: 'Item', isArgument: false, isResult: false, linked: true, name: 'Item', rw: { spec: [Circular] }, strict: false, surface: [Function: thriftrw_Item] }
      actual:
        { rw: { spec: [Circular], valueType: { Constructor: [Function: thriftrw_Item], fieldNames: [ 'key', 'value' ], fields: [ { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: [Object], valueType: [Object] }, { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: [Object], valueType: [Object] } ], fieldsById: { 1: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: [Object], valueType: [Object] }, 2: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: [Object], valueType: [Object] } }, fieldsByName: { key: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 1, name: 'key', optional: false, required: true, valueDefinition: [Object], valueType: [Object] }, value: { annotations: null, constructDefaultValue: null, defaultValue: null, defaultValueDefinition: null, id: 2, name: 'value', optional: false, required: true, valueDefinition: [Object], valueType: [Object] } }, fullName: 'Item', isArgument: false, isResult: false, linked: true, name: 'Item', rw: { spec: [Circular] }, strict: false, surface: [Function: thriftrw_Item] } }, valueType: 'Item' }
      at: Test.t (/home/uber/dsg/thriftrw/test/include.js:47:12)
    ...
```

Comment (malandrew) [inline test/include.js]: I'm guessing that this indicates a bug, but I don't yet understand where/why.

File: test/include.js (hunk 15 above)
Comment (malandrew) [inline test/include.js]: `entryPoint` instead of `thriftFile`? I went with `thriftFile` since that is what I saw used with getThriftSync in the HyperbahnClient. 

Personally, I prefer just `filepath` or `thriftFilepath`. The former I prefer because it seems redundant, when you're doing `Thrift.loadSync`, since Thrift is already in the class name the function is on.

File: test/include.js (hunk 14 above)
Comment (malandrew) [inline test/include.js]: Done!

File: test/include.js (hunk 13 above)
Comment (malandrew) [inline test/include.js]: Resolved.

File: thrift.js (hunk 12 above)
Comment (malandrew) [inline thrift.js]: Resolved!

File: thrift.js (hunk 8 above)
Comment (malandrew) [inline thrift.js]: Resolved!

File: thrift.js (hunk 5 above)
Comment (malandrew) [inline thrift.js]: Error is now thrown. Resolved.

File: thrift.js (hunk 4 above)
Comment (malandrew) [inline thrift.js]: Resolved!

File: thrift.js (hunk 3 above)
Comment (malandrew) [inline thrift.js]: Resolved!

File: thrift.js (hunk 2 above)
Comment (malandrew) [inline thrift.js]: Resolved!

File: thrift.js (hunk 1 above)
Comment (malandrew) [inline thrift.js]: Doesn't seem to make that big a difference per jsPerf, but resolved anyways:

https://jsperf.com/indexof-vs-lastindexof-with-0-start-index

File: thrift.js (hunk 6 above)
Comment (malandrew) [inline thrift.js]: Done!

File: thrift.js (hunk 9 above)
Comment (malandrew) [inline thrift.js]: Done!

File: thrift.js (hunk 11 above)
Comment (malandrew) [inline thrift.js]: Done

File: test/include.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+// THE SOFTWARE.
+
+'use strict';
+
+var test = require('tape');
+
+var Thrift = require('../thrift').Thrift;
+var path = require('path');
+
+test('loads a thrift file with imports synchronously', function t(assert) {
```
Comment (kriskowal) [inline test/include.js]: s/with/which/ s/which/that/

Comment (malandrew) [inline test/include.js]: done

File: thrift.js [hunk 18]
```diff
@@ -46,17 +48,32 @@ var ThriftMap = require('./map').ThriftMap;
 var ThriftConst = require('./const').ThriftConst;
 var ThriftTypedef = require('./typedef').ThriftTypedef;
 
+var validThriftIdentifierRE = /^[a-zA-Z0-9_][a-zA-Z0-9_\.]+$/;
```
Comment (abhinav) [inline thrift.js]: This allows using "42" as an identifier. The first character can only be `[a-zA-Z_]`.

File: thrift.js (hunk 10 above)
Comment (malandrew) [inline thrift.js]: Hidden behind `options.allowIncludeAlias` flag.

File: test/include.js (hunk 16 above)
Comment (malandrew) [inline test/include.js]: I'm almost certain the current behavior is actually correct since this happens with sets, maps and lists. With these three, we want to compare the value type of the element of the set, map or list and  not the valueType of the set/map/list itself

File: thrift.js (hunk 18 above)
Comment (malandrew) [inline thrift.js]: resolved.

File: thrift.js [hunk 19]
```diff
@@ -305,4 +407,26 @@ Thrift.prototype.resolveMapConst = function resolveMapConst(def) {
     return map;
 };
 
+Thrift.prototype.resolveIdentifier = function resolveIdentifier(def, identifier) {
```
Comment (kriskowal) [inline thrift.js]: resolveIdentifierValue (to be distinguished from resolveIdentifier which would return a type)

File: thrift.js [hunk 20]
```diff
@@ -266,21 +376,13 @@ Thrift.prototype.resolveValue = function resolveValue(def) {
...
         }
-        // istanbul ignore if
-        if (!self.consts[def.name]) {
-            err = new Error('cannot resolve reference to ' + def.name + ' at ' + def.line + ':' + def.column);
-            err.line = def.line;
-            err.column = def.column;
-            throw err;
-        }
-        return self.consts[def.name].link(self).surface;
+        return self.resolveIdentifier(def, def.name).surface;
```
Comment (kriskowal) [inline thrift.js]: Just pass def.

File: thrift.js [hunk 21]
```diff
@@ -202,6 +310,19 @@ Thrift.prototype.link = function link() {
...
+        return self;
+    }
+    self.linked = true;
+
+    var moduleNames = Object.keys(self.modulesByName);
+    for (index = 0; index < moduleNames.length; index++) {
+        var thriftModule = self.modulesByName[moduleNames[index]];
+        thriftModule.link(self);
+        self.modulesByName[moduleNames[index]] = thriftModule;
+        self[moduleNames[index]] = thriftModule;
```
Comment (kriskowal) [inline thrift.js]: Let’s not add modules to the thrift object yet. We can use `thrift.modules.LocalModuleName` for now. We don’t have anything in place that would prevent a name collision with Thrift.prototype methods. We probably should fall back to only exposing services on the Thrift object proper since they have PascalCase names. We should also enforce that.

File: thrift.js [hunk 22]
```diff
@@ -99,17 +133,32 @@ Thrift.prototype.baseTypes = {
...
 
-Thrift.prototype.compile = function compile(source) {
+Thrift.prototype.compile = function compile() {
     var self = this;
-    var syntax = idl.parse(source);
+    var syntax = idl.parse(self.source);
     assert.equal(syntax.type, 'Program', 'expected a program');
+    self.compileHeaders(syntax.headers);
     self.compileDefinitions(syntax.definitions);
+    self.compiled = true;
```
Comment (kriskowal) [inline thrift.js]: Let’s remove this since it is no necessary.

File: thrift.js [hunk 23]
```diff
@@ -46,17 +48,32 @@ var ThriftMap = require('./map').ThriftMap;
...
+    }
+
+    if (self.thriftFile && !self.source) {
+        self.source = fs.readFileSync(self.thriftFile, 'ascii');
+    }
 
     self.strict = options.strict !== undefined ? options.strict : true;
 
     self.claims = Object.create(null);
+    self.definitions = Object.create(null);
```
Comment (kriskowal) [inline thrift.js]: This introduces an ambiguity about the meaning of "definition". This appears to be a dictionary of "models" née "specs" and could possibly be rolled up with "claims", merging claim/define.

File: thrift.js [hunk 24]
```diff
@@ -65,14 +82,31 @@ function Thrift(options) {
     self.exceptions = Object.create(null);
     self.unions = Object.create(null);
     self.typedefs = Object.create(null);
+    self.modulesByName = Object.create(null);
+    self.modulesByPath = options.modulesByPath || Object.create(null);
+    self.linked = false;
+    self.filepathThriftMemo = options.filepathThriftMemo || Object.create(null);
```
Comment (kriskowal) [inline thrift.js]: Is it possible to merge modulesByPath with filepathThriftMemo?

File: thrift.js (hunk 20 above)
Comment (malandrew) [inline thrift.js]: you mean `resolveIdentifier`? If so, not possible. There are call sites to this function that don't use the same property on def.

File: thrift.js (hunk 22 above)
Comment (malandrew) [inline thrift.js]: Done.

File: thrift.js [hunk 25]
```diff
@@ -65,14 +82,31 @@ function Thrift(options) {
...
+        self.link();
+    }
 }
 
+Thrift.loadSync = function loadSync(options) {
+    var thrift = new Thrift(options);
+    thrift.compile();
+    thrift.link();
+    return thrift;
+};
```
Comment (kriskowal) [inline thrift.js]: Perfect function.

File: test/include.js (hunk 16 above)
Comment (kriskowal) [inline test/include.js]: I’m convinced. If there’s a fix here, it is a naming fix.

File: thrift.js (hunk 20 above)
Comment (kriskowal) [inline thrift.js]: Cool story.

File: thrift.js (hunk 21 above)
Comment (malandrew) [inline thrift.js]: So we shouldn't copy anything except services?

Comment (kriskowal) [inline thrift.js]: Let’s just not add anything for now. We will probably have to retract everything but Services. Alternately, not expose anything at all.

Comment (malandrew) [inline thrift.js]: I'm in favor of not exposing anything at all.

File: test/include.js [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+test('cyclic dependencies', function t(assert) {
+    var thriftA = Thrift.loadSync({
+        thriftFile: path.join(
+            __dirname,
+            'include-cyclic-a.thrift'
+        ),
+        allowIncludeAlias: true
+    });
+
+    var thriftB = thriftA.modulesByName.value;
```
Comment (kriskowal) [inline test/include.js]: For user experience, let’s rename modulesByName to just modules.

File: thrift.js (hunk 23 above)
Comment (malandrew) [inline thrift.js]: I'm down with merging claims and definitions. I went with definitions since that is what the thrift IDL calls these declarations.

File: test/include-cyclic-a.thrift [hunk 27]
```diff
@@ -0,0 +1,6 @@
+include value "./include-cyclic-b.thrift"
```
Comment (kriskowal) [inline test/include-cyclic-a.thrift]: This will be more clear if the name is `include B "./include-cyclic-b.thrift"`

**Comment by kriskowal** [comment]
I think we should stylistically encourage includes in PascalCase.

File: thrift.js (hunk 24 above)
Comment (malandrew) [inline thrift.js]: Technically yes, except for the latter is more like a cache of all files threaded through and the former is only the modules relevant to the current Thrift instance.

File: test/include.js (hunk 26 above)
Comment (malandrew) [inline test/include.js]: will do

File: thrift.js (hunk 25 above)
Comment (breerly) [inline thrift.js]: Nice.

File: thrift.js (hunk 23 above)
Comment (malandrew) [inline thrift.js]: done

File: thrift.js (hunk 24 above)
Comment (malandrew) [inline thrift.js]: done

File: thrift.js (hunk 25 above)
Comment (malandrew) [inline thrift.js]: coolio

File: test/include.js (hunk 26 above)
Comment (malandrew) [inline test/include.js]: done

File: test/include-cyclic-a.thrift (hunk 27 above)
Comment (malandrew) [inline test/include-cyclic-a.thrift]: Done.

**Comment by kriskowal** [comment]
I’ve added 5 commits. If these look good to you, I believe this is ready to land.
